### ISSUE TITLE
Windows suspend/resume detection

### ISSUE THREAD

**Issue by gyrovague** [AndersMalmgren/FreePIE#105]
It would be nice if FreePIE could subscribe to Windows' suspend/resume notifications, e.g.
https://msdn.microsoft.com/en-us/library/windows/desktop/jj613321(v=vs.85).aspx

and use this to stop a script upon suspend, and restart a script (if it had been running) upon resume (perhaps after some time period, say a few seconds after resume notification).

Not sure whether this affects everybody, but I find my FreePIE scripts silently stop working (no errors logged that I can see) after a Windows suspend/resume, then I need to press shift-F5 and F5  to get them running again. This may be because one or more of the devices referenced by the script perhaps (speculating here) temporarily goes away or something like that.

**Comment by MarijnS95** [comment]
This feature is really helpful, and quite easy to implement. 

@AndersMalmgren This is my first time working with the event system; could you check if I've used it correctly? For instance, is [this](https://github.com/MarijnS95/FreePIE/commit/a1cf68bbf6504c7a7afbf846bf71de083f926e08#diff-61d653d98545cf66ec79b05fa76cae16R56) the right place to set up the event handler? And if so - since it happens in the same class - how about having `PowerModeChanged` call to the handler function directly?

**Comment by AndersMalmgren** [comment]
Shouldn't this be handled in backend and not the front end? The guo does not even need to now the script was ever "stopped"

**Comment by AndersMalmgren** [comment]
*gui

**Comment by AndersMalmgren** [comment]
*know, sorry on a phone :)

**Comment by MarijnS95** [comment]
In `PythonScriptEngine`?

**Comment by AndersMalmgren** [comment]
Yeah, have it suspend and start again from there, seems more correct place of doing it.

**Comment by AndersMalmgren** [comment]
Since it's highly coupled to window power management maybe abstract it thoguh

**Comment by MarijnS95** [comment]
I like the idea of abstract, but I'm running into some problems with the design:

My idea was to make an interface and binding, in such a way that the `PythonScriptEngine` can enable/disable this feature, and that the `SuspendResume` class (which handles power events and should propagate them to the script engine) has access to `IScriptEngine` to start/stop it.
However, for that the script engine needs to expose whether it's running or not, and some way to tell the current filename (so that we can call `Start()` with the correct path) or a function to start the previously used script. Both methods don't sound very neat to me.
Also, having the two bind to eachother creates a cyclic dependency which Ninject doesn't like ;).
What did you have in mind?

To toggle this functionality, I was thinking of adding a toggle function (or are properties available?) to the scope (in `CreateScope`):

            Action<bool> toggleResume = (x) => suspendResume.Enabled = x;
            globals.Add("restartOnResume", toggleResume);

Is that the right way?

**Comment by AndersMalmgren** [comment]
Sorry I don't have the time on my hand to look into this on a detailed level right now, but can't the abstraction just be an singleton started in the bootstrapper that publish events for the anyone to listen to, like the script engine

**Comment by AndersMalmgren** [comment]
You want this for the console version too I guess

**Comment by MarijnS95** [comment]
> but can't the abstraction just be an singleton started in the bootstrapper that publish events for the anyone to listen to, like the script engine

In that case I have to modify the script engine to listen for the messages, which could be a possibility. I could also do it the other way around, where the script engine exposes functionality to add functions to the python scope and publish a related event when called.

> You want this for the console version too I guess

I'm planning on implementing it in `FreePIE.Core` completely, so no need to separately add it there :)

**Comment by gyrovague** [comment]
Any chance of this feature making it into publicly available binary release build anytime soon?

**Comment by AndersMalmgren** [comment]
I need to find someone that can help me with freepie I don't have the time to review all pull requests. Marjin might want to take that role?

**Comment by gyrovague** [comment]
Very much understandable. I'm currently split in many directions myself.

**Comment by MarijnS95** [comment]
Perhaps some, but I'm a bit short on time as well 😅 

That said though, this specific feature request hasn't progressed any further since my last comment - I haven't thought about the implementation yet. If I remember correctly, my intent was to make this a per-script setting, in such a way that the python script can call a function or set a variable to enable/disable it. I had some ideas but I need to look at the code again to remember.
